### ISSUE TITLE
Binary Classification with Single Output Neuron

### ISSUE THREAD

**Issue by kkarrancsu** [Trusted-AI/adversarial-robustness-toolbox#306]
Hello,

1. Does ART support generating adversarial examples on neural networks designed for binary classification with a single output neuron?
2. Are there any examples in PyTorch for how to use a pretrained classifier? All examples I've seen include training using the PyTorchClassifier wrapper. I'm not sure how to instantiate one of these objects if I've already got a trained model.

Thanks!

**Comment by beat-buesser** [comment]
Hi @kkarrancsu Thank you for your interest in ART!

1. The current version of ART v1.1 does not officially support binary classification with 1 output. You would have to use a model which makes predictions for both classes, e.g. with 2 neurons that add up to 1, because we are not testing with binary classifiers and at least a few attacks still contain code that is not compatible with a single output (e.g. `art.utils.get_labels_np_array`, https://github.com/IBM/adversarial-robustness-toolbox/blob/e6d0b53a16bca37712a78ad916018c67a5951abe/art/attacks/evasion/fast_gradient.py#L152, etc.). Binary classification with a single output is definitively something we would like to officially support in future releases, therefore I would like to keep this issue open.

2. Yes, you can use ART with pre-trained classifiers. For example in the `get_started_pytorch.py` (https://github.com/IBM/adversarial-robustness-toolbox/blob/master/examples/get_started_pytorch.py) replace the line 49, which creates a new untrained model with code to load a trained PyTorch model and remove line 63 as you don't need to fit the classifier anymore if the model is already trained. This notebook loads a trained Keras model and creates a KerasClassifier with it, it works similar in PyTorch: https://github.com/IBM/adversarial-robustness-toolbox/blob/master/notebooks/attack_defence_imagenet.ipynb.

Let me know if you have more questions!

**Comment by kkarrancsu** [comment]
Thanks so much for the prompt response @beat-buesser.  Regarding using a trained model, would we still need to wrap our trained model with the `PyTorchClassifier ` class in order to use the ART attacks against it?  In that case, the constructor seems to require both a `torch.nn.modules.loss._Loss` and an `torch.optim.Optimizer` object.  I didn't dig into the code enough to see whether those objects would be unused if the model is already pretrained or not ... if they are unused, I suspect we can just pass `None`, but if not, then how should we proceed?

**Comment by beat-buesser** [comment]
@kkarrancsu Ok, I see. 

Yes, trained models also need to be wrapped with the ART Classifier to provide all the API methods that the attacks are expecting.

The differences in the classifiers have several reasons. I think in Keras it is/was relatively easy to automatically extract the loss function from a model, whereas it seemed more challenging in PyTorch, which is why ART asks the user the define the loss function as input to PyTorchClassifier.

You probably could create and run a PyTorchClassifier with setting `loss` and `optimizer` to None if you only need the classifier's `predict` method which would enable black-box attacks only. For white-box attacks the gradient calculation provided by ART classifier requires the loss function, therefore it will require `loss` to be defined. If you would like to `fit`/train the classifier to new data the `optimizer` is required. My current recommendation would be to define all arguments, if possible.

**Comment by stroypet** [comment]
Hi @beat-buesser ,

Thanks for your ongoing support and creation of ART. What is the current status of this issue? What would it take to allow binary classifiers to work with art?

Cheers

**Comment by beat-buesser** [comment]
Hi @stroypet Thank you very much for using ART!

Unfortunately we didn't make much progress so far because of other projects last year, but there is no doubt that this would be very useful and desired and we'll try to support it this year.

I think it should be possible to integrate it into the existing classifiers in `art.estimators.classification`. The classification estimators currently treat labels internally as one-hot encoded arrays, therefore they would have to check the output of most likely a test prediction with the model to recognise if it is a single output model or not and then adapt their treatment of the labels accordingly. A similar step might be required for the attacks which currently accept labels of shape (N, 1) and treat them as index labels for multi-class classification, they would have to be adapted to check the classifier property from its test prediction to learn if the labels are index labels or binary labels.

I think the good news is that most of the related code that would need updates is in reused functions like `art.utils.check_and_transform_label_format` and testing would reveal other implicit assumptions on label format requiring updates.

Would you be interested to try solving this issue? I would be very happy to provide support.

**Comment by stroypet** [comment]
Hi @beat-buesser 

Sorry for the delayed response. If we (not me, specifically, but my research team), ever free up some cycles in the coming months, we may need to take a stab at fixing this ourselves! We are interested in this functionality. I'm glad you're open for support.

Will officially add it to our backlog... 😄

**Comment by beat-buesser** [comment]
Hi @stroypet 

That's amazing! We just had another issue #902 where support for single output binary classifiers would have been useful.

I'll add this issue to the ART 1.7.0 project (release date would be June 15) to keep track of it but without adding any pressure.

Let me know when you think a Google Hangout might be useful to discuss some of the possible approaches.

**Comment by stroypet** [comment]
@beat-buesser Hi Beat. We are ready to take a deeper look at this now. Would love to get 15 - 30 minutes of your time to chat and figure out this, or any contributions you'd like to see. 

How would you like to link up? I'll send you an email @ beat.buesser@ie.ibm.com and hope it finds you.

**Comment by beat-buesser** [comment]
Hi @stroypet That sounds great!

**Comment by joseph-nagel** [comment]
A little late to the party I suppose. Since the issue still persists, a small workaround is to create an equivalent model with two output neurons. In PyTorch one could for example wrap a model predicting a single logit in another class:
```python
class TwoLogits(nn.Module):

    def __init__(self, model: nn.Module) -> None:
        super().__init__()
        self.model = model

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        y = self.model(x)
        return torch.cat((torch.zeros_like(y), y), dim=-1)
```
This is a simple trick to construct an appropriate two-logits model that is accepted by ART. Of course, it can be done similarly in TensorFLow.
